python/dgl/dataloading/pytorch/dataloader.py

```diff
@@ -0,0 +1,912 @@
+    of message flow graphs (MFGs) as computation dependency of the said minibatch.
+
+    Parameters
+    ----------
+    g : DGLGraph
+        The graph.
+    nids : Tensor or dict[ntype, Tensor]
+        The node set to compute outputs.
+    block_sampler : dgl.dataloading.BlockSampler
+        The neighborhood sampler.
+    device : device context, optional
+        The device of the generated MFGs in each iteration, which should be a
+        PyTorch device object (e.g., ``torch.device``).
+    use_ddp : boolean, optional
+        If True, tells the DataLoader to split the training set for each
+        participating process appropriately using
+        :class:`torch.utils.data.distributed.DistributedSampler`.
+
+        Overrides the :attr:`sampler` argument of :class:`torch.utils.data.DataLoader`.
+    ddp_seed : int, optional
+        The seed for shuffling the dataset in
+        :class:`torch.utils.data.distributed.DistributedSampler`.
+
+        Only effective when :attr:`use_ddp` is True.
+    kwargs : dict
+        Arguments being passed to :py:class:`torch.utils.data.DataLoader`.
+
+    Examples
+    --------
+    To train a 3-layer GNN for node classification on a set of nodes ``train_nid`` on
+    a homogeneous graph where each node takes messages from all neighbors (assume
+    the backend is PyTorch):
+
+    >>> sampler = dgl.dataloading.MultiLayerNeighborSampler([15, 10, 5])
+    >>> dataloader = dgl.dataloading.NodeDataLoader(
+    ...     g, train_nid, sampler,
+    ...     batch_size=1024, shuffle=True, drop_last=False, num_workers=4)
+    >>> for input_nodes, output_nodes, blocks in dataloader:
+    ...     train_on(input_nodes, output_nodes, blocks)
+
+    **Using with Distributed Data Parallel**
+
+    If you are using PyTorch's distributed training (e.g. when using
+    :mod:`torch.nn.parallel.DistributedDataParallel`), you can train the model by turning
+    on the `use_ddp` option:
+
+    >>> sampler = dgl.dataloading.MultiLayerNeighborSampler([15, 10, 5])
+    >>> dataloader = dgl.dataloading.NodeDataLoader(
+    ...     g, train_nid, sampler, use_ddp=True,
+    ...     batch_size=1024, shuffle=True, drop_last=False, num_workers=4)
+    >>> for epoch in range(start_epoch, n_epochs):
+    ...     dataloader.set_epoch(epoch)
+    ...     for input_nodes, output_nodes, blocks in dataloader:
+    ...         train_on(input_nodes, output_nodes, blocks)
+
+    Notes
+    -----
+    Please refer to
+    :doc:`Minibatch Training Tutorials <tutorials/large/L0_neighbor_sampling_overview>`
+    and :ref:`User Guide Section 6 <guide-minibatch>` for usage.
+
+    If you would like to perform GPU-based neighbor sampling, we recommend you
+    
+    * Feed in the graph on GPU.
+    * Set :attr:`device` to GPU.
+    * Set :attr:`num_workers` to 0.
+
+    If your graph is too big, we recommend you
+
+    * Feed in the graph on CPU.
+    * Set :attr:`device` to GPU.
+    * (You can set :attr:`num_workers` to anything you like.)
```

Comment:
Should it be noted here that setting `num_workers > 0` will force `to_block()` to run on the CPU instead of the GPU? Maybe something like:
```
* If :attr:`num_workers` is set to 0 in this case, the sampling will happen on the CPU, and then the subgraphs will be constructed directly on the GPU. This is the recommend setting in multi-GPU configurations.
* Otherwise, if :attr:`num_workers` is greater than 0, both the sampling and subgraph construction will take place on the CPU. This is the recommended setting when using a single-GPU and the whole graph does not fit in GPU memory.
```

Reply:
Let's not use you or we in the sentences. How about:
```
**Tips for selecting the proper device**

* If the input graph :attr:`g` is on GPU, the output device :attr:`device` must be the same GPU
  and :attr:`num_workers` must be zero. In this case, the sampling and subgraph construction
  will take place on the GPU. This is the recommended setting when using a single-GPU and
  the whole graph fits in GPU memory.
* If the input graph :attr:`g` is on CPU while the output device :attr:`device` is GPU, then depending
  on the value of :attr:`num_workers`:
  - If :attr:`num_workers` is set to 0, the sampling will happen on the CPU, and then the subgraphs
    will be constructed directly on the GPU. This is the recommend setting in multi-GPU configurations.
  - Otherwise, if :attr:`num_workers` is greater than 0, both the sampling and subgraph construction
    will take place on the CPU. This is the recommended setting when using a single-GPU and the whole
    graph does not fit in GPU memory.
```

Reply:
Yeah this is much better.